django/core/files/storage/memory.py

```diff
@@ -0,0 +1,243 @@
+
+        return current_node
+
+    def _resolve_child(self, path_segment, create_if_missing, child_cls):
+        if path_segment in ("", "."):
+            return self
+        if create_if_missing:
+            self._update_accessed_time()
+            self._update_modified_time()
+            return self._children.setdefault(path_segment, child_cls())
+        return self._children.get(path_segment)
+
+    def list_dir_names(self):
+        return [k for k, v in self._children.items() if isinstance(v, InMemoryDirNode)]
+
+    def list_file_names(self):
+        return [k for k, v in self._children.items() if isinstance(v, InMemoryFileNode)]
+
+    def remove_child(self, name):
+        if name in self._children:
+            self._update_accessed_time()
+            self._update_modified_time()
+            del self._children[name]
+
+
+@deconstructible(path="django.core.files.storage.InMemoryStorage")
+class InMemoryStorage(Storage, StorageSettingsMixin):
+    """
+    A storage saving files in memory
+    """
+
+    def __init__(self, base_url=None):
+        self._root = InMemoryDirNode()
+        self._base_url = base_url
+        setting_changed.connect(self._clear_cached_properties)
+
+    @cached_property
+    def base_url(self):
+        if self._base_url is not None and not self._base_url.endswith("/"):
+            self._base_url += "/"
+        return self._value_or_setting(self._base_url, settings.MEDIA_URL)
+
+    def _open(self, name, mode="rb"):
+        create_if_missing = "w" in mode
+        file_node = self._root.resolve(
+            name, create_if_missing=create_if_missing, leaf_cls=InMemoryFileNode
+        )
+        return file_node.open(mode)
+
+    def _save(self, name, content):
+        file_node = self._root.resolve(
+            name, create_if_missing=True, leaf_cls=InMemoryFileNode
+        )
+        fd = None
+
+        for chunk in content.chunks():
+            if fd is None:
+                mode = "wb" if isinstance(chunk, bytes) else "wt"
+                fd = file_node.open(mode)
+            fd.write(chunk)
+
+        file_node.modified_time = now()
+        return name
+
+    def delete(self, name):
+        path, filename = os.path.split(name)
+        dir_node = self._root.resolve(path)
+        if dir_node is None:
+            return None
+        dir_node.remove_child(filename)
+
+    def exists(self, name):
+        return self._root.resolve(name, check_exists=False) is not None
+
+    def listdir(self, path):
+        node = self._root.resolve(path, leaf_cls=InMemoryDirNode)
+        return (node.list_dir_names(), node.list_file_names())
+
+    def size(self, name):
+        return len(self._open(name, "rb").file.getvalue())
+
+    def url(self, name):
+        if self.base_url is None:
+            raise ValueError("This file is not accessible via a URL.")
+        url = filepath_to_uri(name)
+        if url is not None:
```

Comment:
Coverage says:
"line 229 didn't jump to line 231, because the condition on line 229 was never false"